### PR TITLE
New Query Pipeline

### PR DESCRIPTION
**Pull request by smitpatel**
What works:
Non-collection queries
Sync/Async
InMemory/Sqlite/SqlServer
Method/Member translators
Spatial

What does not work:
Cosmos provider - Tests are skipped in normal run so will fix later
Collection queries - Includes/Correlated Subquery is not implemented yet
Subquery lifting - Was dependent on nav rewrite
A lot of small optimizations in expression tree/generated Sql
Null semantics
TPH
Broader list at https://github.com/aspnet/EntityFrameworkCore/issues/14630

Notes:
Old files are not deleted yet. (keeping them for reference for now)
Still target NS2.0
Class names/namespaces are not final.
Doc comments missing

### CODE REVIEW

File: src/EFCore.Relational/Query/PipeLine/QuerySqlGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        };
+
+        public QuerySqlGenerator(IRelationalCommandBuilderFactory relationalCommandBuilderFactory,
+            ISqlGenerationHelper sqlGenerationHelper)
+        {
+            _relationalCommandBuilderFactory = relationalCommandBuilderFactory;
+            _sqlGenerationHelper = sqlGenerationHelper;
+        }
+
+        public virtual IRelationalCommand GenerateSql(
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/PipeLine/QuerySqlGenerator.cs]: Might be a good time to rename this to something like 'GetCommand'

Comment (smitpatel) [reply src/EFCore.Relational/Query/PipeLine/QuerySqlGenerator.cs]: There will be huge API review later.

Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/PipeLine/QuerySqlGenerator.cs]: We missed it in the last review, so changing it now will prevent that happening again

---
In reply to: [249137132](https://github.com/aspnet/EntityFrameworkCore/pull/14455#discussion_r249137132) [](ancestors = 249137132)

File: src/EFCore/Metadata/Internal/EntityMaterializerSource.cs [hunk 2]
```diff
@@ -181,7 +182,7 @@ var materializationContextParameter
                         = Expression.Parameter(typeof(MaterializationContext), "materializationContext");
 
                     return Expression.Lambda<Func<MaterializationContext, object>>(
-                            CreateMaterializeExpression(e, materializationContextParameter),
+                            CreateMaterializeExpression(e, "instance", materializationContextParameter),
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/EntityMaterializerSource.cs]: >"instance" [](start = 59, length = 10)

Is this something that would ever change? Should it be just a static readonly field instead?

File: src/EFCore/Query/PipeLine/ShapedQueryExpressionVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                                _materializationContextConstructor,
+                                entityShaperExpression.ValueBufferExpression,
+                                Expression.MakeMemberAccess(
+                                    QueryCompilationContext2.QueryContextParameter,
+                                    _dbContextMemberInfo))));
+
+                    var materializationExpression
+                        = (BlockExpression)_entityMaterializerSource.CreateMaterializeExpression(
+                            entityShaperExpression.EntityType,
+                            "instance" + _currentEntityIndex++,
```
Comment (AndriySvyryd) [inline src/EFCore/Query/PipeLine/ShapedQueryExpressionVisitor.cs]: >"instance" [](start = 28, length = 10)

Use predefined name?

File: src/EFCore/Query/PipeLine/ShapedQueryExpressionVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+            private int _currentEntityIndex;
+
+
+            public EntityMaterializerInjectingExpressionVisitor(IEntityMaterializerSource entityMaterializerSource)
+            {
+                _entityMaterializerSource = entityMaterializerSource;
+            }
+
+            protected override Expression VisitExtension(Expression extensionExpresssion)
```
Comment (AndriySvyryd) [inline src/EFCore/Query/PipeLine/ShapedQueryExpressionVisitor.cs]: Typo in 'extensionExpresssion'

File: src/EFCore/Metadata/Internal/EntityMaterializerSource.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore/Metadata/Internal/EntityMaterializerSource.cs]: Not from this call site. Other call site passes different name. Hence the method started taking string param.

File: src/EFCore/Query/PipeLine/ShapedQueryExpressionVisitor.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore/Query/PipeLine/ShapedQueryExpressionVisitor.cs]: Still playing around with materializer. This code will change. Will incorporate predefined names once it is finalized.

**Comment by benmccallum** [comment]
https://github.com/aspnet/EntityFrameworkCore/pull/14971 cross-linking so I can see over there easily when this PR is merged.

File: src/EFCore/Query/PipeLine/QueryOptimizingExpressionVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        private static SelectManyVerifyingExpressionVisitor _selectManyVerifyingExpressionVisitor
+            = new SelectManyVerifyingExpressionVisitor();
+        private static EnumerableToQueryableReMappingExpressionVisitor _enumerableToQueryableReMappingExpressionVisitor
+            = new EnumerableToQueryableReMappingExpressionVisitor();
+
+        protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
+        {
+            if (methodCallExpression.Method.DeclaringType == typeof(Queryable))
+            {
+                if (methodCallExpression.Method.IsGenericMethod)
```
Comment (maumar) [inline src/EFCore/Query/PipeLine/QueryOptimizingExpressionVisitor.cs]: combine those two ifs

**Comment by smitpatel** [comment]
This is ready for review. 
I am in process of disabling failing tests (or fix them if easy win). Once this is in, it would be easier to work in smaller chunks

File: src/EFCore/Query/PipeLine/QueryOptimizingExpressionVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+                    return parameterExpression;
+                }
+
+                _correlated = true;
+
+                return base.VisitParameter(parameterExpression);
+            }
+        }
+
+        private LambdaExpression UnwrapLambdaFromQuoteExpression(Expression expression)
```
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore/Query/PipeLine/QueryOptimizingExpressionVisitor.cs]: Will to DRY later

File: src/EFCore.InMemory/Query/PipeLine/Translator.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+using System.Reflection;
+using Microsoft.EntityFrameworkCore.Extensions.Internal;
+using Microsoft.EntityFrameworkCore.Query;
+using Microsoft.EntityFrameworkCore.Query.Expressions.Internal;
+using Microsoft.EntityFrameworkCore.Query.Internal;
+using Microsoft.EntityFrameworkCore.Query.Pipeline;
+
+namespace Microsoft.EntityFrameworkCore.InMemory.Query.Pipeline
+{
+    public class InMemoryExpressionTranslatingExpressionVisitor : ExpressionVisitor
```
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore.InMemory/Query/PipeLine/Translator.cs]: Will cover in API review

File: src/EFCore.Relational/Query/PipeLine/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 8]
```diff
@@ -0,0 +1,872 @@
...
+            }
+
+            throw new NotImplementedException();
+        }
+
+        protected override ShapedQueryExpression TranslateCount(ShapedQueryExpression source, LambdaExpression predicate)
+        {
+            var selectExpression = (SelectExpression)source.QueryExpression;
+
+            if (selectExpression.IsDistinct
```
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore.Relational/Query/PipeLine/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: Not all conditions are same. I intend to dry them once they are finalized. Don't want unintentional push down by mistake from DRY code.

File: src/EFCore.Relational/Query/PipeLine/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 9]
```diff
@@ -0,0 +1,872 @@
...
+
+            var selectExpression = (SelectExpression)source.QueryExpression;
+
+            if (selectExpression.Limit != null
+                || selectExpression.Offset != null)
+            {
+                selectExpression.PushdownIntoSubQuery();
+            }
+
+            selectExpression.Reverse();
```
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore.Relational/Query/PipeLine/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: https://github.com/aspnet/EntityFrameworkCore/issues/15559

File: src/EFCore.Relational/Query/PipeLine/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,872 @@
...
+            {
+                selectExpression.PushdownIntoSubQuery();
+            }
+
+            if (predicate != null)
+            {
+                source = TranslateWhere(source, predicate);
+            }
+
+            var translation = _sqlExpressionFactory.ApplyDefaultTypeMapping(
```
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore.Relational/Query/PipeLine/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: Will add later when adding SqlServer specific Sql Gen.

File: src/EFCore/Query/PipeLine/QueryOptimizingExpressionVisitor.cs [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+            return base.VisitMethodCall(methodCallExpression);
+        }
+
+        private class EnumerableToQueryableReMappingExpressionVisitor : ExpressionVisitor
+        {
+            protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
+            {
+                if (methodCallExpression.Method.DeclaringType == typeof(Enumerable))
+                {
+                    var queryableMethod = typeof(Queryable).GetTypeInfo().GetDeclaredMethods(methodCallExpression.Method.Name)
```
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore/Query/PipeLine/QueryOptimizingExpressionVisitor.cs]: https://github.com/aspnet/EntityFrameworkCore/issues/15560

**Comment by smitpatel** [comment]
Addressed feedback and disabled all failing tests. Build should be passing.

**Comment by maumar** [comment]
:shipit: 
design looks good, we will clean it up and iron out the kinks in small chunks

File: src/EFCore/Query/PipeLine/EntityQueryableExpressionVisitor2.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using Microsoft.EntityFrameworkCore.Internal;
+
+namespace Microsoft.EntityFrameworkCore.Query.Pipeline
```
Comment (AndriySvyryd) [inline src/EFCore/Query/PipeLine/EntityQueryableExpressionVisitor2.cs]: Why the folder is `PipeLine` and not `Pipeline`? :sob:

Comment (smitpatel) [reply src/EFCore/Query/PipeLine/EntityQueryableExpressionVisitor2.cs]: Rename went bad. I will correct it to please unicorns.

Comment (roji) [reply src/EFCore/Query/PipeLine/EntityQueryableExpressionVisitor2.cs]: Yes this was a major issue hurting my eyes as well.

**Pull request merged**